File: py/grid.py
from random import randrange, shuffle

from const import chars


class Grid:
    def __init__(self, rows=10, obstacles=5):
        self.grid = []
        self.rows = rows
        self.obstacles = obstacles
        self.max_obstacles = int(self.rows ** 2 / 2)
        self._create()

    def _fill_grid(self):
        for x in range(self.rows):
            self.grid.append([])
            self.grid[x] = [""] * self.rows

    def _obstacles(self):
        if self.obstacles > self.max_obstacles:
            print("Too many obstacles:", self.obstacles, f"[max {self.max_obstacles}]")
            self.obstacles = self.max_obstacles

        for n in range(self.obstacles):
            x = randrange(self.rows) if n >= self.rows else n

            while self.grid[x][x] or self.grid[x].count(chars["obstacle"]) > self.rows:
                x = randrange(self.rows)

            self.grid[x][x] = chars["obstacle"]

            shuffle(self.grid)

    def _create(self):
        self._fill_grid()
        self._obstacles()

File: js/grid.js
class Grid {
  constructor(rows = 10, obstacles = 5) {
    this.grid = [];
    this.rows = rows;
    this.obstacles = obstacles;
    this.max_obstacles = parseInt(this.rows ** 2 / 2);
    this.create();
  }

  fillGrid() {
    for (let i = 0; i < this.rows; i++) {
      const row = Array(this.rows).fill('');
      this.grid.push(row);
    }
  }

  shuffleGrid() {
    const grid = this.grid;
    for (let i = grid.length - 1; i > 0; i--) {
      const j = Math.floor(Math.random() * (i + 1));
      [grid[i], grid[j]] = [grid[j], grid[i]];
    }
    return grid;
  }

  createObstacles() {
    if (this.obstacles > this.max_obstacles) {
      console.log(
          'Too many obstacles:', this.obstacles,
          `[max ${this.max_obstacles}]`,
      );

      this.obstacles = this.max_obstacles;
    }

    for (let i = 0; i < this.obstacles; i++) {
      let x = i >= this.rows ? Math.floor(Math.random() * this.rows) : i;

      const count = this.grid[x].filter((e) => e === obsChar).length;

      while (this.grid[x][x] || count > this.rows) {
        x = Math.floor(Math.random() * this.rows);
      }

      this.grid[x][x] = obsChar;

      this.shuffleGrid();
    }
  }

  create() {
    this.fillGrid();
    this.createObstacles();
  }
};
